From 8416d6b7f4cd296e38cbed705887a91ccd4022a6 Mon Sep 17 00:00:00 2001
From: Sora Su <sora_su@wiwynn.com>
Date: Thu, 2 Apr 2026 09:56:55 +0800
Subject: [PATCH 18/22] oem-meta: Implement special event handling

Certain events must be recorded by directly invoking phosphor-logging.
In particular, the CXL HB requires controlling the event severity
based on its current state.

Tested: Verified on the Yosemite4 platform. All events were recorded
correctly.

Change-Id: I5bec004d181122ae7bac7aec347ea8692a5dc9fb
Signed-off-by: Sora Su <sora_su@wiwynn.com>
---
 .../file_io_type_event_log.cpp                | 85 +++++++++++++++++--
 1 file changed, 80 insertions(+), 5 deletions(-)

diff --git a/oem/meta/libpldmresponder/file_io_type_event_log.cpp b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
index 45902cb..261b627 100644
--- a/oem/meta/libpldmresponder/file_io_type_event_log.cpp
+++ b/oem/meta/libpldmresponder/file_io_type_event_log.cpp
@@ -724,6 +724,81 @@ static void rainbowFault(const message& eventData, const std::string&,
     recordEventLog(updatedFault, eventStatus, eventData[3], eventData[4]);
 }
 
+static void cxlEvent(const message& eventData, const std::string& event,
+                     const std::string&)
+{
+    lg2::info("{INFO}", "INFO", event); // Create log in journal
+    auto& bus = pldm::utils::DBusHandler::getBus();
+    using EntrySeverity =
+        sdbusplus::xyz::openbmc_project::Logging::server::Entry::Level;
+
+    static constexpr auto loggingService = "xyz.openbmc_project.Logging";
+    static constexpr auto loggingPath = "/xyz/openbmc_project/logging";
+    static constexpr auto loggingInterface =
+        "xyz.openbmc_project.Logging.Create";
+
+    std::string severityStr;
+    EventAssert eventStatus = static_cast<EventAssert>(eventData[1]);
+    if (eventStatus == EventAssert::DEASSERTED)
+    {
+        severityStr =
+            sdbusplus::xyz::openbmc_project::Logging::server::convertForMessage(
+                EntrySeverity::Error);
+    }
+    else
+    {
+        severityStr =
+            sdbusplus::xyz::openbmc_project::Logging::server::convertForMessage(
+                EntrySeverity::Informational);
+    }
+
+    try
+    {
+        auto method = bus.new_method_call(loggingService, loggingPath,
+                                          loggingInterface, "Create");
+
+        std::map<std::string, std::string> addlData{};
+        method.append(event.c_str(), severityStr, addlData);
+        bus.call_noreply(method, dbusTimeout);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Failed to create info-level D-Bus log: {ERR}", "ERR", e);
+    }
+}
+
+static void platformEvent([[maybe_unused]] const message& eventData,
+                          const std::string& event, const std::string&)
+{
+    lg2::info("{INFO}", "INFO", event); // Create log in journal
+    auto& bus = pldm::utils::DBusHandler::getBus();
+    using EntrySeverity =
+        sdbusplus::xyz::openbmc_project::Logging::server::Entry::Level;
+
+    static constexpr auto loggingService = "xyz.openbmc_project.Logging";
+    static constexpr auto loggingPath = "/xyz/openbmc_project/logging";
+    static constexpr auto loggingInterface =
+        "xyz.openbmc_project.Logging.Create";
+
+    auto severityStr =
+        sdbusplus::xyz::openbmc_project::Logging::server::convertForMessage(
+            EntrySeverity::Informational);
+
+    try
+    {
+        auto method = bus.new_method_call(loggingService, loggingPath,
+                                          loggingInterface, "Create");
+
+        std::map<std::string, std::string> addlData{};
+        method.append(event.c_str(), severityStr, addlData);
+        bus.call_noreply(method, dbusTimeout);
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Failed to create info-level D-Bus log: {ERR}", "ERR", e);
+    }
+}
+
 static void reportError(const message&, const std::string& event,
                         const std::string&)
 {
@@ -805,11 +880,11 @@ static const std::map<EventType, EventDescriptor> eventHandlers = {
     {EventType::POWER_ON_SEQUENCE_FAILURE, {report::powerRailFault, nullptr}},
     {EventType::DIMM_PMIC_ERROR,
      {report::powerRailFault, format::dimmPmicError}},
-    {EventType::CXL1_HB, {nullptr, nullptr}},
-    {EventType::CXL2_HB, {nullptr, nullptr}},
-    {EventType::PLTRST_ASSERTION, {nullptr, nullptr}},
-    {EventType::POST_STARTED, {nullptr, nullptr}},
-    {EventType::POST_ENDED, {nullptr, nullptr}},
+    {EventType::CXL1_HB, {report::cxlEvent, nullptr}},
+    {EventType::CXL2_HB, {report::cxlEvent, nullptr}},
+    {EventType::PLTRST_ASSERTION, {report::platformEvent, nullptr}},
+    {EventType::POST_STARTED, {report::platformEvent, nullptr}},
+    {EventType::POST_ENDED, {report::platformEvent, nullptr}},
     {EventType::PROCHOT_IS_TRIGGERED_DUE_TO_SD_SENSOR_READING_EXCEED_UCR,
      {report::prochotSdSensor, nullptr}},
     {EventType::MTIA_FAULT, {report::mtiaFault, nullptr}},
-- 
2.52.0

